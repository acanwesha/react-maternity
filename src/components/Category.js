import React, { useEffect, useState } from 'react'
import { Collapse, ListGroup, ListGroupItem } from "react-bootstrap"

const Category = props => {
  const {
    category,
    categoryIndex,
    selectedCategory,
    setSelectedCategory,
    methods,
    finalCategoryArray,
  } = props
  const { setCategoryArray } = methods

  const [collapsedState, setCollapsedState] = useState(
    categoryIndex === selectedCategory //  ===-1
  )
  const [subCheckbox, setSubCheckbox] = useState(false)

  useEffect(() => {
    setCollapsedState(categoryIndex === selectedCategory)
  }, [categoryIndex, selectedCategory])

  return (
    <div>
      <ListGroup>
        <span>
          <ListGroupItem style={{ backgroundColor: "#cceeff" }}>
            <div>
              <input
                type='checkbox'
                name='categoryId'
                data-name="category-checkbox"
                // id={category.id}
                onChange={e => {
                  setSubCheckbox(!subCheckbox)
                  const displayName = category.displayName
                  if (e.target.checked) {
                    //  if selected then added new category
                    const updatedArray = [
                      ...finalCategoryArray,
                      { displayName: displayName },
                    ]
                    setCategoryArray(updatedArray)
                    //                    console.log("added category", updatedArray)
                  } else {
                    //  if unselected then remove category from array
                    const updatedArray = finalCategoryArray.filter(
                      obj => obj.displayName !== displayName
                    )
                    setCategoryArray(updatedArray)
                    //                  console.log("removed category", updatedArray)
                  }
                }}
              />
              <label
                //    htmlFor={category.id}
                data-testid="displayName"
                style={{ cursor: "pointer" }}
                onClick={() => {
                  setSelectedCategory(categoryIndex)
                  setCollapsedState(!collapsedState)
                }}
                className='ml-2'
              >
                {category.displayName}
              </label>
            </div>
            <Collapse
              in={collapsedState}
              style={{
                position: "absolute",
                left: "348px",
                top: "0px",
                width: "450px",
                backgroundColor: "#dcf9fa"
              }}
            >
              <div id={category.id}>
                {category.result.map((item, index) => {
                  if (Object.keys(item).length === 0) {
                    return ""
                  }
                  //  if we unselect the main category then also need to unselect the sub categories
                  if (!subCheckbox) {
                    var checkboxElement = document.getElementById(item.id)
  
                    if (checkboxElement !== null) {
                      if (checkboxElement.checked === true) {
                        checkboxElement.checked = false
                      }
                    }
                  }
                  return (
                    <ListGroup 
                      key={index} 
                      className='ml-1'
                      data-name="item"
                    >
                      <span>
                        <ListGroupItem style={{ backgroundColor: "#caecfc" }}>
                          <div> 
                            <input
                              type='checkbox'
                              name='itemId'
                              data-name="sub-category-checkbox"
                              id={item.id}
                              disabled={!subCheckbox}
                              onChange={e => {
                                const displayName = item.displayName

                                //  first get the category index
                                const currentCategoryId =
                                  finalCategoryArray.findIndex(
                                    object =>
                                      object.displayName === category.displayName
                                  )
                                //  create a duplicate array for modification
                                const duplicateArray = [...finalCategoryArray]
                                var resultArray = []
                                if (
                                  duplicateArray[currentCategoryId].result !==
                                  undefined
                                ) {
                                  resultArray = [
                                    ...duplicateArray[currentCategoryId].result,
                                  ]
                                }
                                if (e.target.checked) {
                                  //  added into array
                                  const resultObject = {
                                    displayName: displayName,
                                  }
                                  if (item.value !== undefined) {
                                    resultObject["value"] = item.value
                                  }
                                  resultArray.push(resultObject)
                                  duplicateArray[currentCategoryId].result =
                                    resultArray
                                  setCategoryArray(duplicateArray)
                                   console.log(
                                    "sub-category added ",
                                    duplicateArray
                                  )
                                } else {
                                  // removed from array
                                  duplicateArray[currentCategoryId].result =
                                    resultArray.filter(
                                      object => object.displayName !== item.displayName
                                    )

                                  setCategoryArray(duplicateArray)
                                  console.log(
                                    "sub-category removed ",
                                    duplicateArray
                                  )
                                }
                              }}
                            />
                            <label
                              htmlFor={item.id}
                              style={{ cursor: "pointer" }}
                              className='ml-2'
                            >
                              {item.displayName}
                            </label>
                            {item.value !== undefined && (
                              <span>
                                : <b> {item.value}</b>
                              </span>
                            )}
                          </div>
                        </ListGroupItem>
                      </span>
                    </ListGroup>
                  )
                })}
              </div>
            </Collapse>
          </ListGroupItem>
        </span>
      </ListGroup>
    </div>
  )
}
export default Category;
