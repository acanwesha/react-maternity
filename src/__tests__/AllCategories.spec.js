import { shallow } from 'enzyme';
import AllCategory from '../components/AllCategories';
import Category from '../components/Category';


describe('components/AllCategories', () => {
 let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<AllCategory {...props} />)
  }

  beforeEach(() => {
    props = {
      data: [
        {
          id: 1
        },
        {
          id: 2
        }
      ], methods: {}, finalCategoryArray: []
    }
  })

  it('should renders correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when no categories available', () => {
    props.data = [];
    setup();
    expect(wrapper).toMatchSnapshot();
  })

  it('should handle setting selectedCategory correctly', () => {
    setup();
    expect(wrapper.find(Category).first().props().selectedCategory).toEqual(-1)
    wrapper.find(Category).first().props().setSelectedCategory(1);
    expect(wrapper.find(Category).first().props().selectedCategory).toEqual(1)
  })
})
