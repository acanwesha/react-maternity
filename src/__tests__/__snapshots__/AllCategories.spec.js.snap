// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AllCategories should render correctly when no categories available 1`] = `
<div>
  <h3>
    Category
  </h3>
  <p />
  No categories available
</div>
`;

exports[`components/AllCategories should renders correctly 1`] = `
<div>
  <h3>
    Category
  </h3>
  <p />
  <Category
    category={
      Object {
        "id": 1,
      }
    }
    categoryIndex={0}
    finalCategoryArray={Array []}
    key="0"
    methods={Object {}}
    selectedCategory={-1}
    setSelectedCategory={[Function]}
  />
  <Category
    category={
      Object {
        "id": 2,
      }
    }
    categoryIndex={1}
    finalCategoryArray={Array []}
    key="1"
    methods={Object {}}
    selectedCategory={-1}
    setSelectedCategory={[Function]}
  />
</div>
`;
